migrate.go

```diff
@@ -389,34 +395,74 @@ func PlanMigration(db *sql.DB, dialect string, m MigrationSource, dir MigrationD
 	return result, dbMap, nil
 }
 
-// Filter a slice of migrations into ones that should be applied.
-func ToApply(migrations []*Migration, current string, direction MigrationDirection) []*Migration {
-	var index = -1
-	if current != "" {
-		for index < len(migrations)-1 {
-			index++
-			if migrations[index].Id == current {
+// Searches the arrays for the latest elements with common ids.
+// Returns the indexes of the common id.
+func lastCommon(left []*Migration, right []*Migration) (int, int) {
+	xIndexMatch := -1
+	yIndexMatch := 0
+	for i := 0; i < len(left); i++ {
+		existingId := left[i].Id
+		for j := yIndexMatch; j < len(right); j++ {
+			if right[j].Id == existingId {
+				xIndexMatch = i
+				yIndexMatch = j
 				break
 			}
 		}
 	}
+	if xIndexMatch == -1 {
+		return -1, -1 // We never found a match; the arrays have nothing in common
+	}
+	return xIndexMatch, yIndexMatch
+}
 
+// Filter a slice of migrations into ones that should be applied.
+func ToApply(migrations, existingMigrations []*Migration, direction MigrationDirection) []*Migration {
 	if direction == Up {
-		return migrations[index+1:]
-	} else if direction == Down {
-		if index == -1 {
-			return []*Migration{}
-		}
+		return toApplyUp(migrations, existingMigrations)
+	}
+	if direction == Down {
+		return toApplyDown(migrations, existingMigrations)
+	}
+	panic("Unknown direction")
+}
 
-		// Add in reverse order
-		toApply := make([]*Migration, index+1)
-		for i := 0; i < index+1; i++ {
-			toApply[index-i] = migrations[i]
-		}
-		return toApply
+func toApplyUp(migrations, existingMigrations []*Migration) []*Migration {
+	if len(existingMigrations) == 0 {
+		return migrations
+	}
+	_, index := lastCommon(existingMigrations, migrations)
+	return migrations[index+1:]
+}
+
+func toApplyDown(migrations, existingMigrations []*Migration) []*Migration {
+	if len(existingMigrations) == -1 {
+		return []*Migration{}
+	}
+	// When a Migration appears in both existingMigrations and migrations,
+	// we want the Migration from migrations, since only that list has
+	// the Up and Down fields set.
```

Comment:
Would this comment make more sense if it were on the second loop, which does the overwriting you describe?